Speed space chart: several improvements of the reticle, steps and crosshair

flomonster opened the pull request:
Main changes :
1. Gather steps rectangle, lines and title together
2. Match style with mock-up (color, size, opacity)
3. Vertical line can be visible without step name
4. Step snapping matches the mock-up
5. Add vetical and horizontal lines from crosshair to axis

Fixes:
- Steps with duplicate positions are now draw only once.
- Fix reticle shown position
- Fix snapped step position
- Snapped steps have their text hidden

![image](https://github.com/user-attachments/assets/cec9df56-838b-40bd-a7d7-c7008359de7f)

clarani commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +266,49 @@ export const createSvgBlobUrl = (svgString: string): string => {
+  MARGINS.MARGIN_LEFT +
+  leftOffset;
+
+/** Interpole on a straight line giving two points and the x-axis value */
```
```suggestion
/** Interpole on a straight line given two points and the x-axis value */
```

clarani commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +266,49 @@ export const createSvgBlobUrl = (svgString: string): string => {
+    ((width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES) * ratioX) +
+  MARGINS.CURVE_MARGIN_SIDES / 2 +
+  MARGINS.MARGIN_LEFT +
+  leftOffset;
```
Maybe it would be easier to read like this
```suggestion
  const xWidth = width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES;
  const leftMargin = MARGINS.CURVE_MARGIN_SIDES / 2 + MARGINS.MARGIN_LEFT + leftOffset;
  return (position / maxPosition) * (xWidth * ratioX) + leftMargin;
```

Yohh commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +266,49 @@ export const createSvgBlobUrl = (svgString: string): string => {
   width: number,
   ratioX: number,
-  maxPosition: number
+  leftOffset: number = 0
```
```suggestion
  leftOffset = 0
```

Yohh commented on ui-speedspacechart/src/components/helpers/drawElements/gridX.ts:
```diff
@@ -1,63 +1,114 @@
+  ratioX: number,
+  width: number,
+  maxPosition: number,
+  offset: number = 8
```
```suggestion
  offset = 8
```

Yohh commented on ui-speedspacechart/src/components/layers/GridX.tsx:
```diff
@@ -4,13 +4,13 @@ import type { Store } from '../../types/chartTypes';
 };
 
-const AxisLayerX = ({ width, height, store }: AxisLayerXProps) => {
+const GridX = ({ width, height, store }: GridXProps) => {
```
`GridX.tsx` contains lines and steps, we can keep `AxisLayerX.tsx`

flomonster replied:
It doesn't contains the ticks. It feels weird to named it axis since it's more a layer over the x axis.

Yohh replied:
I still have a problem with that:
- every layers have the word `layer` in their name, except `GridX.tsx`
- `TickLayerY.tsx` contains `drawAxisY` that draws ticks, text, minor and major lines.
- `TickLayerX.tsx` contains `drawTickX` that draws only ticks and texts.
- `GridX.tsx` contains `drawGridX` that draws lines, steps and texts.

we're losing clarity, we should keep consistent names, maybe add folders for axis logic to keep an optimal readability and navigation

flomonster replied:
I suggest: `StepsLayer`.

- It's closer to what this layer represents
- It contains the word `Layer`

It makes a lot of sense when we look at this piece of code:

```ts
      {store.layersDisplay.steps && (
        <StepsLayer width={adjustedWidthRightAxis} height={height} store={store} />
      )}
```

> [!NOTE]
> I'm not a big fan of the name "step" everywhere in osrd this object is named "operationnal point" (or "op"). We can rename this later.

Yohh replied:
I named it steps because even if this is obvious for us, I'm no sure this is really relevant for anyone outside of the project how could use this module

Yohh commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -18,6 +21,49 @@ const {
+  return Math.min(Math.max(0, cursorPosition), maxPosition);
+};
+
+export const getSnappedStop = (cursorX: number, width: number, store: Store) => {
```
should be in `utils.ts` and tested

Yohh reviewed (commented):
nice! thanks for this

need more time to properly review the reticle

`getStopPosition` doesn't pass in tests check

Yohh commented:
are the two horiszontal lines on the cruves pointing to the left axis refined and confirmed by @thibautsailly ?

![image](https://github.com/user-attachments/assets/b8c1de7b-03df-4e79-870a-c9fdec006e58)

flomonster commented:
> are the two horizontal lines on the curves pointing to the left axis refined and confirmed by @thibautsailly ?

It's part of the mock-up:

![image](https://github.com/user-attachments/assets/c8a26f4d-4f26-4f67-b601-dab24de034f2)

Yohh commented:
> > are the two horizontal lines on the curves pointing to the left axis refined and confirmed by @thibautsailly ?
> 
> It's part of the mock-up:
> 
> ![image](https://private-user-images.githubusercontent.com/15349030/367316843-c8a26f4d-4f26-4f67-b601-dab24de034f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyMzUwNzgsIm5iZiI6MTcyNjIzNDc3OCwicGF0aCI6Ii8xNTM0OTAzMC8zNjczMTY4NDMtYzhhMjZmNGQtNGYyNi00ZjY3LWI2MDEtZGFiMjRkZTAzNGYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEzVDEzMzkzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlMzhkNDk5MDRiZTY0MTM1MmJkMGY0NDJiN2Q4ODkzZTg5MGIzMTE3YmE2ZGU2ZGIwMjE5ODg0OGZjZmY3ZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dGZkmxx9uE8FLMXpuyKo5wIdyyt6kuHr4cSFAZYL8EY)

yes I saw it, but the mockup is versioned, so what we can see at first view could change at anytime.

flomonster committed:
speedspacechart: speed-up and rewrite utils functions

- Use binary search instead of linear

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

flomonster committed:
speedspacechart: rewrite steps and reticle management

- Gather steps rectangle, lines and title together
  - Hidden if the layer is disabled
- Match style with mock-up (color, size, opacity)
- Step snapping matches the mock-up
- Add vetical and horizontal lines from crosshair to axis
- Multiple small fixes

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

flomonster commented:
I've moved part of this PR into another ([here](https://github.com/OpenRailAssociation/osrd-ui/pull/480)) for easier review.

jacomyal commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +267,133 @@ export const createSvgBlobUrl = (svgString: string): string => {
+  ratioX: number,
+  width: number,
+  maxPosition: number,
+  offset = 8
```
I think `offset` is a confusing name here, as it does not describe an offset, but the minimal space between two consecutive displayed stops. May be `minSpace` would be better.

jacomyal commented on ui-speedspacechart/src/components/common/DetailsBox.tsx:
They were already there before this PR, but there are a lot of "magic numbers" in that file, and some of them look like they depend on each other. It might be better to store them in constant variables, with names that explain what these numbers correspond to on screen.

flomonster replied:
I reduced the number of magic numbers. 

Is it okay to have magic number as proportion of a constant size?
Example: `SNAPPED_CURSOR_HEIGHT * 0.6`

jacomyal replied:
OK for me. At least the dependence and the name are clear.

jacomyal commented on ui-speedspacechart/src/components/helpers/drawElements/gridX.ts:
There are now three different colors hard-coded within the rendering code. It would be nice to have them in constant variables instead, with names describing what they correspond to precisely.

flomonster replied:
I reduced the number of magic numbers. 

Is it okay to have magic number as proportion of a constant size?
Example: `SNAPPED_CURSOR_HEIGHT * 0.6`

flomonster replied:
Fixed. I didn't know how to handle different level of opacity for a same color properly.

I'm crafting colors using format string to add opacity value. I don't think it's very readable. Any advice?

Yohh replied:
we don't have so many levels of opacity, I think they can be set in their own color variable.

in the other hand we could have variables for colors and variables for opacity levels, then concatenate them if we need instead of having magic-numbers-like in strings, this will be easier if we keep hexa.

jacomyal replied:
> Fixed. I didn't know how to handle different level of opacity for a same color properly.
> 
> I'm crafting colors using format string to add opacity value. I don't think it's very readable. Any advice?

Personally, I would use the "small-ish" JS library dedicated to colors management [chroma-js](https://www.npmjs.com/package/chroma-js) (already a dependency of the `spacetimechart` by the way). It would do something like `const transparentColor = chroma(color).alpha(0.5).rgba()` for instance.

flomonster replied:
Awesome, exactly what I was looking for :pray:

jacomyal commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -79,225 +70,179 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
-          cursor.x! <= xPositionReference(position.end!)
-      );
-      powerRestrictionText = currentPowerRestriction?.value.powerRestriction || '';
+  // Check if cursor is snappping to stop
```
Typo (snapp**p**ing)

jacomyal commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -79,225 +70,179 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
-        electricalProfileText = electricalProfileValue;
-      }
-    }
+  let cursorPosition = getCursorPosition(cursor.x!, width, store);
```
The `!` is not necessary here, because of the previous early escape on `cursor.x === null`. Normally, TypeScript should "understands" that here `cursor.x` is necessarily a number. This comment also applies for the next occurrences of `cursor.x!` in this function.

flomonster replied:
Done. TypeScript is not able to detect that the value not null when used in the lambda function. Check [here](https://github.com/OpenRailAssociation/osrd-ui/pull/455/files#diff-60f8e6d4b91b5be2ea8787dbc4ae1dcb69c6d99d9f6876a6a1fb8c90833b47dbR99-R100)

thibautsailly commented:
> are the two horizontal lines on the curves pointing to the left axis refined and confirmed by @thibautsailly ?
> 
> ![image](https://private-user-images.githubusercontent.com/20640252/367252822-b8c1de7b-03df-4e79-870a-c9fdec006e58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0OTQ2MTIsIm5iZiI6MTcyNjQ5NDMxMiwicGF0aCI6Ii8yMDY0MDI1Mi8zNjcyNTI4MjItYjhjMWRlN2ItMDNkZi00ZTc5LTg3MGEtYzlmZGVjMDA2ZTU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDEzNDUxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3MDAwYjRkNzFlNzM2Y2MwMzljNDEyMDM4ZTM0YzczODI3NTk2NzY4MWFkZDc2NTdjMWFhOWI3NjkzNzI3NmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5GtxlolN-7Gfsx_k5qS284Kni0kaYjbRVCMWgkSGDBI)

Confirmed

Caracol3 commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -79,225 +73,185 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
+  );
+  if (predecessorEcoSpeedIndex === ecoSpeeds.length - 1) predecessorEcoSpeedIndex--;
+
+  const speed1 = speeds[predecessorSpeedIndex];
```
Maybe rename those two const for more clarity. 
'prevSpeed' and 'nextSpeed' ?

Caracol3 commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -79,225 +73,185 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
+  const baseReticleY =
+    cursorBoxHeight - (speedValue / maxSpeed) * (cursorBoxHeight - CURVE_MARGIN_TOP) + MARGIN_TOP;
+
+  const ecoSpeed1 = ecoSpeeds[predecessorEcoSpeedIndex];
```
Same as above.

Yohh commented on ui-speedspacechart/src/components/const.ts:
```diff
@@ -63,6 +65,13 @@ export const LAYERS_SELECTION: Array<keyof Store['layersDisplay']> = [
 
+// Colors
+
+export const BLACK = '#000000';
+export const GREY_50 = '#797671';
+export const GREY_80 = '#312e2b';
+export const LIGHT_BLUE = '#2170b9';
```
we usually use rgba() instead of hexa in OSRD-UI, is there a reason to use hexa in this pr?

flomonster replied:
I took example from the space time chart. 

Kind of related. Do you have a suggestion for this issue: https://github.com/OpenRailAssociation/osrd-ui/pull/455/files/9fab2503afcb3d252edc2304a4f9a2754cfaf383#r1761472603 ?

Caracol3 commented:
Is it a known issue ?
In the example, when zooming in, we see Vanves-Malakoff BI popping but when zooming in again it disappears and is replaced by Vanves-Malakoff 05. It appears again with another zoom level.

https://github.com/user-attachments/assets/815e85bf-0373-4b2c-a8be-42fbd2ab0e40

flomonster commented:
> Is it a known issue ? In the example, when zooming in, we see Vanves-Malakoff BI popping but when zooming in again it disappears and is replaced by Vanves-Malakoff 05. It appears again with another zoom level.
> Enregistrement.de.l.ecran.2024-09-17.a.15.36.30.mov

I've noticed this issue. It's due to the naivety of the function that selects which stops overlap or not. 
This  behaviour is not introduced by this PR (since I reused the same strategy as previously). I suggest to fix this in another PR if prioritized.

Yohh commented:
> > Is it a known issue ? In the example, when zooming in, we see Vanves-Malakoff BI popping but when zooming in again it disappears and is replaced by Vanves-Malakoff 05. It appears again with another zoom level.
> > Enregistrement.de.l.ecran.2024-09-17.a.15.36.30.mov
> 
> I've noticed this issue. It's due to the naivety of the function that selects which stops overlap or not. This behaviour is not introduced by this PR (since I reused the same strategy as previously). I suggest to fix this in another PR if prioritized.

this is nor just about the way the text is display while zooming, now the texts and the prs are displayed in their own way each, wich seems to be a new unwanted behavior, but yes this should be fixed in another issue.

flomonster commented:
> basically the steps are displayed if they have enough space to not overlap (16px), this is why it can change at each zoom level, 

This is why it can change at each zoom level you mean. 

Example, if you take an overlap of 16 pixels and you have steps at **A**: 0px,  **B**: 15px and **C**: 16px.
- At a normal zoom level, **A** and **C** will be displayed
- If you zoom a little, **A** and **B** will be displayed (and **C** will disapear)

That's why we had the problem before.

> but this is not just about the way the text is display while zooming, now the texts and the prs are displayed in their own way each, which seems to be a new unwanted behavior, but yes this should be fixed in another issue.

It's not an unwanted behavior at all. They should be displayed in their own way given the mock-up.
I agree that not correcting the zoom stability and adding this behavior. The two combined give a strange effect.

Caracol3 reviewed (approved):
lgtm

Yohh commented:
just one question that doesn't have to block this PR, do we really need to install a whole lib (chromajs) to use it at last only for 3 cases?

flomonster commented:
> just one question that doesn't have to block this PR, do we really need to install a whole lib (chroma-js) to use it at last only for 3 cases?
> 
> edit: To be more precise: we install an entire node module, which we will have to maintain, while 3 variables seem to be enough

We should refacto completely how colors are used across osrd-ui. We should have a library of defined colors so each component doesn't have to redefined their own colors. Having a library here is okay for readability. Using hexa codes was really bad. 
I could declare all colors with their opacity level but the naming would have been unclear. How do you name the color `GREY_50` with an opacity level of `0.5`?

clarani commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +267,133 @@ export const createSvgBlobUrl = (svgString: string): string => {
+) => {
+  let lastDisplayedPosX: number | null = null;
+
+  const filteredStops = stops.filter(({ position }) => {
```
Here you suppose that the stops are sorted by position.start right ?

flomonster replied:
Yes

clarani commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +267,133 @@ export const createSvgBlobUrl = (svgString: string): string => {
+  return filteredStops;
+};
+
+/** Compute the cursor position on the x-axis given it's position on the canva */
```
```suggestion
/** Compute the cursor position on the x-axis given its position on the canva */
```

clarani commented on ui-speedspacechart/src/components/utils.ts:
```diff
@@ -266,24 +267,133 @@ export const createSvgBlobUrl = (svgString: string): string => {
+  return (
+    ((cursorX - leftOffset - MARGINS.CURVE_MARGIN_SIDES / 2) * maxPosition) /
+    ((width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES) * ratioX)
+  );
```
```suggestion
  const x = cursorX - leftOffset - MARGINS.CURVE_MARGIN_SIDES / 2;
  const maxX =
    (width - MARGINS.MARGIN_LEFT - MARGINS.MARGIN_RIGHT - MARGINS.CURVE_MARGIN_SIDES) * ratioX;
  return (x * maxPosition) / maxX;
```

clarani commented on ui-speedspacechart/src/__tests__/utils.spec.ts:
```diff
@@ -351,8 +348,49 @@ describe('getStopPosition', () => {
     const maxPosition = 600;
-    const stopPosition = getStopPosition(position, width, ratioX, maxPosition);
-    expect(stopPosition).toBe(336);
+    const stopPosition = positionToPosX(position.start, maxPosition, width, ratioX);
```
You forgot to change the name of the tested function: getStopPosition => positionToPosX
```suggestion
    const x = positionToPosX(position.start, maxPosition, width, ratioX);
```

clarani commented on ui-speedspacechart/src/components/layers/StepsLayer.tsx:
```diff
@@ -1,17 +1,17 @@
 import { useCanvas } from '../hooks';
 
-type AxisLayerXProps = {
+type GridXProps = {
```
```suggestion
type StepsLayerProps = {
```

clarani commented on ui-speedspacechart/src/components/layers/StepsLayer.tsx:
I think the id of the layer should be `"steps-layer"` and not `"axis-layer-x"`

clarani commented on ui-speedspacechart/src/components/helpers/drawElements/steps.ts:
```diff
@@ -0,0 +1,104 @@
+  ctx.beginPath();
+  filteredStops.forEach(({ position }) => {
+    const posX = positionToPosX(position.start, maxPosition, width, ratioX);
+    ctx.moveTo(posX, height - (height + MARGIN_TOP));
```
```suggestion
    ctx.moveTo(posX, MARGIN_TOP);
```

clarani commented on ui-speedspacechart/src/components/helpers/drawElements/steps.ts:
```diff
@@ -0,0 +1,104 @@
+  ctx.strokeStyle = GREY_50.hex();
+  ctx.setLineDash([]);
+
+  const snappedStop = cursor.x ? getSnappedStop(cursor.x!, width, store) : null;
```
```suggestion
  const snappedStop = cursor.x ? getSnappedStop(cursor.x, width, store) : null;
```

clarani commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -56,18 +58,10 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
-    a: 0,
-    b: 0,
-  };
+  let reticleX = cursor.x! + MARGIN_LEFT;
```
Move the lines 76-79 above this block and remove the !
```suggestion
  let reticleX = cursor.x + MARGIN_LEFT;
```

clarani commented on ui-speedspacechart/src/components/helpers/drawElements/reticle.ts:
```diff
@@ -79,225 +73,185 @@ export const drawCursor = ({ ctx, width, height, store }: DrawFunctionParams) =>
-      powerRestrictionText = currentPowerRestriction?.value.powerRestriction || '';
+  // Check if cursor is snapping to stop
+  let snapToStop = false;
+  if (store.layersDisplay.steps) {
```
```suggestion
  if (layersDisplay.steps) {
```

clarani reviewed (approved):
LGTM :white_check_mark:
